test/widgets/login_test.dart

```diff
@@ -26,19 +37,112 @@ void main() {
+    expectErrorFromText('!@#*asd;l4fkj', ServerUrlValidationError.invalidUrl);
+    expectErrorFromText(
+        'email@example.com', ServerUrlValidationError.noUseEmail);
+  });
+
+  // TODO test AddAccountPage
+
+  group('PasswordLoginPage', () {
+    late FakeApiConnection connection;
+
+    Future<void> prepare(
+        WidgetTester tester, GetServerSettingsResult serverSettings) async {
+      addTearDown(testBinding.reset);
+
+      connection = testBinding.globalStore.apiConnection(
+          realmUrl: serverSettings.realmUrl,
+          zulipFeatureLevel: serverSettings.zulipFeatureLevel);
+
+      await tester.pumpWidget(MaterialApp(
+          localizationsDelegates: ZulipLocalizations.localizationsDelegates,
+          supportedLocales: ZulipLocalizations.supportedLocales,
+          home: GlobalStoreWidget(
+              child: PasswordLoginPage(serverSettings: serverSettings))));
+      await tester.pump(); // load global store
+    }
+
+    final findUsernameInput = find.byWidgetPredicate((widget) =>
+        widget is TextField &&
+        (widget.autofillHints ?? []).contains(AutofillHints.email));
+    final findPasswordInput = find.byWidgetPredicate((widget) =>
+        widget is TextField &&
+        (widget.autofillHints ?? []).contains(AutofillHints.password));
+    final findSubmitButton = find.widgetWithText(ElevatedButton, 'Log in');
+
+    void checkFetchApiKey(
+        {required String username, required String password}) {
+      check(connection.lastRequest).isA<http.Request>()
+        ..method.equals('POST')
+        ..url.path.equals('/api/v1/fetch_api_key')
+        ..bodyFields.deepEquals({
+          'username': username,
+          'password': password,
+        });
+    }
+
+    testWidgets('basic happy case', (tester) async {
+      final serverSettings = eg.serverSettings();
+      await prepare(tester, serverSettings);
+      check(testBinding.globalStore.accounts).isEmpty();
+
+      await tester.enterText(findUsernameInput, eg.selfAccount.email);
+      await tester.enterText(findPasswordInput, 'p455w0rd');
+      connection.prepare(
+          json: FetchApiKeyResult(
+        apiKey: eg.selfAccount.apiKey,
+        email: eg.selfAccount.email,
+        userId: eg.selfAccount.userId,
+      ).toJson());
+      await tester.tap(findSubmitButton);
+      checkFetchApiKey(username: eg.selfAccount.email, password: 'p455w0rd');
+      await tester.idle();
+      check(testBinding.globalStore.accounts).single.equals(eg.selfAccount
+          .copyWith(id: testBinding.globalStore.accounts.single.id));
+    });
+
+    testWidgets('test case with trimming logic', (tester) async {
+      final serverSettings = eg.serverSettings();
+      await prepare(tester, serverSettings);
+      check(testBinding.globalStore.accounts).isEmpty();
+
+      final emailWithTrailingSpaces = '${eg.selfAccount.email}    ';
+      await tester.enterText(findUsernameInput, emailWithTrailingSpaces);
```

Comment:
May as well add some spaces at the beginning as well as the end, to exercise both sides of what `String.trim` does.